core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java

```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
             }
         }
     }
+
+    @Override
+    public AggregatorFactories readFrom(StreamInput in) throws IOException {
+        int factoriesSize = in.readVInt();
+        List<AggregatorFactory> factoriesList = new ArrayList<AggregatorFactory>(factoriesSize);
+        for (int i = 0; i < factoriesSize; i++) {
+            AggregatorFactory factory = in.readAggregatorFactory();
+            factoriesList.add(factory);
+        }
+        int pipelineFactoriesSize = in.readVInt();
+        List<PipelineAggregatorFactory> pipelineAggregatorFactoriesList = new ArrayList<PipelineAggregatorFactory>(pipelineFactoriesSize);
+        for (int i = 0; i < pipelineFactoriesSize; i++) {
+            PipelineAggregatorFactory factory = in.readPipelineAggregatorFactory();
+            pipelineAggregatorFactoriesList.add(factory);
+        }
+        AggregatorFactories aggregatorFactories = new AggregatorFactories(
+                factoriesList.toArray(new AggregatorFactory[factoriesList.size()]), pipelineAggregatorFactoriesList);
+        boolean hasParent = in.readBoolean();
+        if (hasParent) {
+            AggregatorFactory parent = in.readAggregatorFactory();
+            aggregatorFactories.parent = parent;
+        }
+        return aggregatorFactories;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(this.factories.length);
+        for (AggregatorFactory factory : factories) {
+            out.writeAggregatorFactory(factory);
+        }
+        out.writeVInt(this.pipelineAggregatorFactories.size());
+        for (PipelineAggregatorFactory factory : pipelineAggregatorFactories) {
+            out.writePipelineAggregatorFactory(factory);
+        }
+        boolean hasParent = parent != null;
+        out.writeBoolean(hasParent);
+        if (hasParent) {
+            out.writeAggregatorFactory(parent);
```

Comment:
I suspect there is an issue with this. Imagine we have a two-level aggregation, then the top-level aggs will be written twice, once on its own and once as a parent of its sub aggregations?